netebpfext/net_ebpf_ext.c

```diff
@@ -806,6 +856,80 @@ net_ebpf_extension_uninitialize_wfp_components(void)
                 NET_EBPF_EXT_TRACELOG_KEYWORD_EXTENSION, "FwpsInjectionHandleDestroy", status);
         }
     }
+
+    // If there are cleanup filters, sleep to give WFP time to issue callbacks. Once this timeout completes,
+    // we assume that any remaining notifications will not be issued, and proceed with cleanup.
+    KIRQL old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_wfp_cleanup_state.lock);
+    if (!IsListEmpty(&_net_ebpf_ext_wfp_cleanup_state.filter_cleanup_list)) {
+        NET_EBPF_EXT_LOG_MESSAGE(
+            NET_EBPF_EXT_TRACELOG_LEVEL_VERBOSE,
+            NET_EBPF_EXT_TRACELOG_KEYWORD_EXTENSION,
+            "Leaked WFP Filters found. Processing cleanup.");
+        _net_ebpf_ext_wfp_cleanup_state.signal_empty_filter_list = TRUE;
+
+        // Allow some time for WFP to signal deletion for the filters.
+        ExReleaseSpinLockExclusive(&_net_ebpf_ext_wfp_cleanup_state.lock, old_irql);
+        status = KeWaitForSingleObject(
+            &_net_ebpf_ext_wfp_cleanup_state.wfp_filter_cleanup_event, Executive, KernelMode, FALSE, &timeout);
+        old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_wfp_cleanup_state.lock);
+
+#pragma warning(push)
+#pragma warning(disable : 6001) // Using uninitialized memory 'filter_context' and 'filter_id'.
+        // Proceed with cleanup - assume that any remaining notifications will never be issued by WFP,
+        // and continue with cleaning up our internal state.
+        while (!IsListEmpty(&_net_ebpf_ext_wfp_cleanup_state.filter_cleanup_list)) {
+            uint32_t leaked_filter_count = 0;
+            PLIST_ENTRY entry = RemoveHeadList(&_net_ebpf_ext_wfp_cleanup_state.filter_cleanup_list);
+            net_ebpf_extension_wfp_filter_context_t* filter_context =
+                CONTAINING_RECORD(entry, net_ebpf_extension_wfp_filter_context_t, link);
+            filter_context->in_cleanup_list = FALSE;
+
+            // Release lock as we process the entry. DEREFERENCE_FILTER_CONTEXT also acquires the lock.
+            ExReleaseSpinLockExclusive(&_net_ebpf_ext_wfp_cleanup_state.lock, old_irql);
+
+            // Anything left in the NET_EBPF_EXT_WFP_FILTER_DELETING is considered leaked. Remove the references
+            // to allow for cleanup.
+            for (index = 0; index < filter_context->filter_ids_count; index++) {
+                net_ebpf_ext_wfp_filter_id_t* filter_id = &filter_context->filter_ids[index];
+                if (filter_id->state == NET_EBPF_EXT_WFP_FILTER_DELETING) {
+                    leaked_filter_count++;
+                    NET_EBPF_EXT_LOG_MESSAGE_UINT64(
+                        NET_EBPF_EXT_TRACELOG_LEVEL_VERBOSE,
```

Comment:
Should this be an error or warning level? Right now, it's considered unlikely for WFP to call back beyond 10 seconds, but the timing may change.